datarobot_batch_scoring/network/network.py

```diff
@@ -27,7 +28,45 @@
 
 logger = logging.getLogger(__name__)
 FakeResponse = collections.namedtuple('FakeResponse', 'status_code, text')
+lock = threading.Lock()
 
+try:
+    from functools import lru_cache
+except ImportError:
+    # Python 2.7 compatible lru_cache implementation
+    # https://stackoverflow.com/questions/17119154/python-decorator-optional-argument
+    import functools
+    def lru_cache(*setting_args, **setting_kwargs):
+        cache = {}
+        no_args = (
+            len(setting_args) == 1 and
+            not setting_kwargs and
+            callable(setting_args[0])
+        )
+        if no_args:
+            # We were called without args
+            func = setting_args[0]
+        def outer(func):
+            @functools.wraps(func)
+            def cacher(*args, **kwargs):
+                with lock:
+                    key = tuple(args) + tuple(sorted(kwargs.items()))
+                    if key not in cache:
+                        cache[key] = func(*args, **kwargs)
+                    return cache[key]
+            return cacher
+        return outer(func) if no_args else outer
+
+
+# Monkey-patch getaddrinfo with an LRU cache to minimize conflicting calls
+from requests.packages.urllib3.util.connection import socket as r_socket
+old_getaddrinfo = r_socket.getaddrinfo
+
+@lru_cache()
+def my_getaddrinfo(*args, **kwargs):
+    return old_getaddrinfo(*args, **kwargs)
```

Comment:
I think it would be better to put `lock` here instead of putting it inside a decorator. First, there're code branches in python3 builtin `lru_cache` that are not covered with lock (maxsize is 0, or None). Second, it would be a little bit more explicit that we are really interested in acquiring lock here. What do you think?

Reply:
We don't need lock if we are not calling getaddrinfo. So it's absolutely acceptable and preferable to put it inside of decorator

Reply:
3.6 have thread-safe lru_cache, so we are just matching 3.6 behaviour